webapp/channels/src/components/root/root_redirect/root_redirect.test.tsx

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {createMemoryHistory} from 'history';
+import React from 'react';
+import type {RouteComponentProps} from 'react-router-dom';
+import {Redirect, useHistory} from 'react-router-dom';
+
+import {getFirstAdminSetupComplete as getFirstAdminSetupCompleteAction} from 'mattermost-redux/actions/general';
+
+import * as GlobalActions from 'actions/global_actions';
+
+import {renderWithContext, waitFor} from 'tests/react_testing_utils';
+
+import RootRedirect from './root_redirect';
+import type {Props} from './root_redirect';
+
+jest.mock('actions/global_actions', () => ({
+    redirectUserToDefaultTeam: jest.fn(),
+}));
+
+jest.mock('mattermost-redux/actions/general', () => ({
+    getFirstAdminSetupComplete: jest.fn(() =>
+        Promise.resolve({
+            data: true,
+        }),
+    ),
+}));
+
+jest.mock('react-router-dom', () => {
+    const actual = jest.requireActual('react-router-dom');
+    return {
+        ...actual,
+        Redirect: jest.fn(() => null),
+        useHistory: jest.fn(),
+    };
+});
+
+describe('components/RootRedirect', () => {
+    const baseProps: Props = {
+        currentUserId: '',
+        isElegibleForFirstAdmingOnboarding: false,
+        isFirstAdmin: false,
+        areThereTeams: false,
+        actions: {
+            getFirstAdminSetupComplete: getFirstAdminSetupCompleteAction as jest.Mock,
+        },
+    };
+
+    const defaultProps = {
+        ...baseProps,
+        location: {
+            pathname: '/',
+        },
+    } as Props & RouteComponentProps;
+
+    afterEach(() => {
+        jest.clearAllMocks();
+    });
+
+    test('should redirect to /login when currentUserId is empty', () => {
+        renderWithContext(<RootRedirect {...defaultProps}/>);
+
+        expect(Redirect).toHaveBeenCalledTimes(1);
+        expect(Redirect).toHaveBeenCalledWith(
+            expect.objectContaining({
+                to: expect.objectContaining({
+                    pathname: '/login',
+                }),
+            }),
+            {},
+        );
+    });
+
+    test('should call GlobalActions.redirectUserToDefaultTeam when user is logged in and not eligible for first admin onboarding', () => {
+        const props = {
+            ...defaultProps,
+            currentUserId: 'test-user-id',
+            isElegibleForFirstAdmingOnboarding: false,
+        };
+
+        renderWithContext(<RootRedirect {...props}/>);
+
+        expect(GlobalActions.redirectUserToDefaultTeam).toHaveBeenCalledTimes(1);
+    });
+
+    test('should redirect to preparing-workspace when eligible for first admin onboarding and no teams created', async () => {
+        const history = createMemoryHistory({initialEntries: ['/']});
+        const mockHistoryPush = jest.fn();
+        (useHistory as jest.Mock).mockReturnValue({
+            push: mockHistoryPush,
+        });
```

Comment:
Instead of needing to mock `useHistory`, I think you might be able to use `jest.spyOn` to do this
```suggestion
        const history = createMemoryHistory({initialEntries: ['/']});
        const mockHistoryPush = jest.spyOn(history, 'push');
```
I'm hoping that that might let use `react-router-dom` without any mocking so that the test is as real as possible